### PR TITLE
117 extend date validation with moment

### PR DESCRIPTION
**Pull request by mikeerickson**
- [x] Refactored `date` rule to perform additional test using `moment` if installed
- [x] Updated `spec/date-rule.js` to conditionally handle bad date testing
- [x] Ran full test suite to confirm passing and dist built

NOTE: `moment` lib is included for `dev` mode but listed as `peerDependency` for distribution purposes.   Moment is NOT REQUIRED to use Validator, but will validate more accurately when installed.

### CODE REVIEW

File: spec/date-rule.js [hunk 1]
```diff
@@ -39,14 +39,18 @@ describe('date rule', function() {
...
         date: assert
       }, {
         date: 'date'
       });
-      expect(validator.passes()).to.be.false;
-      expect(validator.fails()).to.be.true;
+      if (idx === 0) {
+        expect(validator.passes()).to.be.true;
+      }
+      if ((idx === 1) || (idx === 2) || (idx === 3)) {
```
Comment (iamdtang) [inline spec/date-rule.js]: I think if we're going to check against individual dates in `asserts`, we should get rid of the `forEach` and write out each of those assertions for clarity.

File: src/rules.js [hunk 2]
```diff
@@ -277,10 +277,19 @@ var rules = {
...
-    if (typeof val === 'number') {
-      return val.toString().length === 12 && valid;
+  date: function(val, format) {
+    var moment;
+    var valid = false;
+    try {
+      moment = require('moment');
+      valid = moment(new Date(val), format, true).isValid();
+    } catch (e) {
+      if(e.code ==='MODULE_NOT_FOUND') {
```
Comment (iamdtang) [inline src/rules.js]: Is the condition that checks for `e.code` necessary?

**Review by iamdtang — CHANGES_REQUESTED** [review]

File: spec/date-rule.js (hunk 1 above)
Comment (mikeerickson) [reply spec/date-rule.js]: @skaterdav85 works for me, i will make adjustment

File: src/rules.js (hunk 2 above)
Comment (mikeerickson) [reply src/rules.js]: @skaterdav85 I don't see any compelling reason for making the check... I will remove it.

File: spec/date-rule.js [hunk 3]
```diff
@@ -32,21 +32,14 @@ describe('date rule', function() {
   });
 
   it('should fail for incorrect date format', function() {
-    var asserts = [
```
Comment (iamdtang) [inline spec/date-rule.js]: I think we still want tests for each item in `asserts`

**Review by iamdtang — CHANGES_REQUESTED** [review]

File: spec/date-rule.js (hunk 3 above)
Comment (mikeerickson) [reply spec/date-rule.js]: @skaterdav85 Would you prefer the it block to have 4 different assertions, or would you prefer for it blocks, with one assertion each?

Comment (iamdtang) [reply spec/date-rule.js]: I'd say each one will be its own test with 1 assertion

**Comment by mikeerickson** [comment]
@skaterdav85 hey David, I made the requested changes.  Perhaps you misssed it.

File: src/rules.js [hunk 4]
```diff
@@ -277,10 +277,17 @@ var rules = {
...
 
-  date: function(val) {
-    var valid = (new Date(val).toString()) !== 'Invalid Date';
-    if (typeof val === 'number') {
-      return val.toString().length === 12 && valid;
+  date: function(val, format) {
+    var moment;
+    var valid = false;
+    try {
+      moment = require('moment');
```
Comment (iamdtang) [inline src/rules.js:284]: hey so will this work in the browser, or will it throw an error like "undefined (require) is not a function"?

**Comment by iamdtang** [comment]
@mikeerickson hey added one more comment

**Comment by mikeerickson** [comment]
@skaterdav85 Ya, you are indeed correct... It was passing here because my build had moment bundled into the UMD module (which it should not have).

So, I have spent the morning creating a separate validation routine, which I am going to suggest as a replacement for moment (if you dont mind)

**Comment by mikeerickson** [comment]
Hey David

If you don’t mind, can you add me as user to repo so that I can assign myself to any issue that I could handle.   I will basically try and triage all tickets and fix them as they come up


Regards,

Mike Erickson

**Comment by iamdtang** [comment]
@mikeerickson cool! are you basically going create a lighter weight moment for the specific cases for this library?

**Comment by mikeerickson** [comment]
@skaterdav85 Yes, that is the plan :-)

**Pull request closed without merging**